.github/workflows/codegen.yml

```diff
@@ -0,0 +1,53 @@
+on:
+  schedule:
+    # daily at 12:30 am
+    - cron: '30 0 * * *'
+  workflow_dispatch:
+
+name: codegen
+jobs:
+  discovery:
+    uses: googleapis/discovery-artifact-manager/.github/workflows/list-services.yml@master
+
+  generate:
+    runs-on: ubuntu-latest
+    needs: discovery
+    strategy:
+      fail-fast: false
+      max-parallel: 4
+      matrix: 
+        service: ${{fromJson(needs.discovery.outputs.services)}}
+    steps:
+      - run: echo generating ${{ matrix.service }}
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 1
+          path: google-api-php-client-services
+      - uses: actions/checkout@v2
+        with:
+          repository: googleapis/discovery-artifact-manager
+          fetch-depth: 1
+          path: discovery-artifact-manager
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 2.7
+      - id: find-preferred-versions
+        run: |
+          versions=$( cat discovery-artifact-manager/discoveries/index.json | jq -r '[.items[] | select(.name == "${{ matrix.service }}" and .name != "discovery" and .name != "websecurityscanner") | select(.name == "admin" or .preferred).version] | join(",")' )
```

Comment:
Can we either clean this logic up to make it more readable, or define the logic elsewhere? 

Also, I don't think this really needs to be here... can't we ensure "discovery" and "websecurityscanner" don't show up in `matrix.service` instead? 

```
select(.name == "${{ matrix.service }}" and .name != "discovery" and .name != "websecurityscanner") |
```

If not, we can add a different step to exclude them, such as:

```
if: ${{ matrix.service != "discovery" and matrix.service != "websecurityscanner" }} 
```

Reply:
I can't update the initial list (it's a shared external workflow that lists all the available services), but I should be able to exclude separately via a [matrix exclusion](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-configurations-from-a-matrix) or a separate `if` clause.